Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.cpp

```diff
@@ -0,0 +1,437 @@
+    (*counts)[res] += 1;
+  }
+
+  return res;
+}
+
+}  // namespace
+void addMolToNetwork(const ROMol &mol, ScaffoldNetwork &network,
+                     const ScaffoldNetworkParams &params) {
+  auto ismi = MolToSmiles(mol);
+  boost::shared_ptr<ROMol> fmol(flattenMol(mol, params));
+  if (params.pruneBeforeFragmenting) {
+    boost::shared_ptr<ROMol> pmol(pruneMol(*fmol, params));
+    fmol.swap(pmol);
+  }
+  auto fsmi = MolToSmiles(*fmol);
+  if (ismi != fsmi) {
+    auto iidx = addEntryIfMissing(network.nodes, ismi, &network.counts);
+    auto fidx = addEntryIfMissing(network.nodes, fsmi, &network.counts);
+    addEntryIfMissing(network.edges,
+                      NetworkEdge({iidx, fidx, EdgeType::Initialize}));
+
+    if (params.includeGenericScaffolds) {
+      bool doAtoms = true;
+      bool doBonds = false;
+      std::unique_ptr<ROMol> gmol(makeScaffoldGeneric(*fmol, doAtoms, doBonds));
+      auto gsmi = MolToSmiles(*gmol);
+      auto gidx = addEntryIfMissing(network.nodes, gsmi, &network.counts);
+      addEntryIfMissing(network.edges,
+                        NetworkEdge({fidx, gidx, EdgeType::Generic}));
+      if (params.includeGenericBondScaffolds) {
+        bool doAtoms = true;
+        bool doBonds = true;
+        std::unique_ptr<ROMol> gbmol(
+            makeScaffoldGeneric(*fmol, doAtoms, doBonds));
+        auto gbsmi = MolToSmiles(*gbmol);
+        auto gbidx = addEntryIfMissing(network.nodes, gbsmi, &network.counts);
+        addEntryIfMissing(network.edges,
+                          NetworkEdge({fidx, gbidx, EdgeType::GenericBond}));
+      }
+    }
+  } else {
+    // add the base molecule to the network
+    addEntryIfMissing(network.nodes, fsmi, &network.counts);
+  }
+
+  auto frags = getMolFragments(*fmol, params);
+  for (auto frag : frags) {
+    auto smi = frag.first;
+    ROMOL_SPTR fragMol = frag.second;
+
+    // the first node is certain to already be in the network, so don't
+    // update counts for it
+    auto iidx = addEntryIfMissing(network.nodes, smi);
+    auto lsmi = MolToSmiles(*fragMol);
+    auto lidx = addEntryIfMissing(network.nodes, lsmi, &network.counts);
+    addEntryIfMissing(network.edges,
+                      NetworkEdge({iidx, lidx, EdgeType::Fragment}));
+
+    if (params.includeGenericScaffolds) {
+      bool doAtoms = true;
+      bool doBonds = false;
+      std::unique_ptr<ROMol> gmol(
+          makeScaffoldGeneric(*fragMol, doAtoms, doBonds));
+      auto gsmi = MolToSmiles(*gmol);
+      auto gidx = addEntryIfMissing(network.nodes, gsmi, &network.counts);
+      addEntryIfMissing(network.edges,
+                        NetworkEdge({lidx, gidx, EdgeType::Generic}));
+      if (params.includeScaffoldsWithAttachments &&
+          params.includeScaffoldsWithoutAttachments) {
+        std::unique_ptr<ROMol> amol(removeAttachmentPoints(*gmol, params));
+        auto asmi = MolToSmiles(*amol);
+        auto aidx = addEntryIfMissing(network.nodes, asmi, &network.counts);
```

Comment:
This add to nodes and edges is such a common operation, shouldn't they be combined to one function?

Reply:
Given that it's a single, relatively simple, templated function for both, I think that would actually end up complicating things.